gravitee-am-test/specs/gateway/mfa/mfa-general.spec.ts

```diff
@@ -0,0 +1,392 @@
+    const loginPage = await withRetry(() => followUpGet(authResponse, 200));
+
+    let xsrf = await extractDomValue(loginPage, '[name=X-XSRF-TOKEN]');
+    let action = await extractDomAttr(loginPage, 'form', 'action');
+
+    const loginPostResponse = await withRetry(() =>
+      performFormPost(
+        action,
+        '',
+        {
+          'X-XSRF-TOKEN': xsrf,
+          username: testContext.domain.user.username,
+          password: testContext.domain.user.password,
+          rememberMe: 'off',
+          client_id: testContext.application.clientId,
+        },
+        {
+          Cookie: loginPage.headers['set-cookie'],
+          'Content-type': 'application/x-www-form-urlencoded',
+        },
+      ).expect(302),
+    );
+
+    const authResponseFinal = await withRetry(() => followUpGet(loginPostResponse, 302));
+
+    expect(authResponseFinal.headers['location']).toBeDefined();
+    expect(authResponseFinal.headers['location']).toContain(testContext.application.redirectUris[0]);
+    expect(authResponseFinal.headers['location']).toContain('code=');
+  });
+});
+
+describe('With one enabled factor should omit MFA flow', () => {
+  it('when stepUp=false', async () => {
+    const applicationSettings = {
+      ...defaultApplicationSettings(),
+      settings: {
+        mfa: {
+          factor: {
+            defaultFactorId: testContext.domain.factors[0].id,
+            applicationFactors: [testContext.domain.factors[0]],
+          },
+          stepUpAuthenticationRule: '{{ false }}',
+          stepUpAuthentication: { active: true, stepUpAuthenticationRule: '{{ false }}' },
+        },
+      },
+    };
+
+    await patchApplication(testContext.domain.domainId, testContext.admin.accessToken, applicationSettings, testContext.application.id);
+
+    const authResponse = await withRetry(() => get(testContext.oidc.clientAuthorizationEndpoint, 302));
+    const loginPage = await withRetry(() => followUpGet(authResponse, 200));
+
+    let xsrf = await extractDomValue(loginPage, '[name=X-XSRF-TOKEN]');
+    let action = await extractDomAttr(loginPage, 'form', 'action');
+
+    const loginPostResponse = await withRetry(() =>
+      performFormPost(
+        action,
+        '',
+        {
+          'X-XSRF-TOKEN': xsrf,
+          username: testContext.domain.user.username,
+          password: testContext.domain.user.password,
+          rememberMe: 'off',
+          client_id: testContext.application.clientId,
+        },
+        {
+          Cookie: loginPage.headers['set-cookie'],
+          'Content-type': 'application/x-www-form-urlencoded',
+        },
+      ).expect(302),
+    );
+
+    const authResponseFinal = await withRetry(() => followUpGet(loginPostResponse, 302));
+
+    expect(authResponseFinal.headers['location']).toBeDefined();
+    expect(authResponseFinal.headers['location']).toContain(testContext.application.redirectUris[0]);
+    expect(authResponseFinal.headers['location']).toContain('code=');
```

Comment:
**TODO**: here we are testing step Up, stepUp should be evaluate only on active session. So to make this test valid, after line 166, we should do another call to `clientAuthorizationEndpoint` and check that we directly receives a code

Reply:
fixed